contracts/amm/SirenExchange.sol

```diff
@@ -0,0 +1,280 @@
+        );
+
+        TransferHelper.safeApprove(
+            path[path.length - 1],
+            sirenAmmAddress,
+            collateralPremium
+        );
+
+        //Call MinterAmm bTokenBuy contract
+        ISirenTradeAMM(sirenAmmAddress).bTokenBuy(
+            seriesId,
+            bTokenAmount,
+            collateralPremium
+        );
+
+        //Transfer the btokens to the correct address ( caller of this contract)
+        erc1155Controller.safeTransferFrom(
+            address(this),
+            msg.sender,
+            SeriesLibrary.bTokenIndex(seriesId),
+            bTokenAmount,
+            dataReturn()
+        );
+
+        emit BTokenBuy(
+            amounts,
+            path,
+            sirenAmmAddress,
+            bTokenAmount,
+            seriesId,
+            msg.sender
+        );
+
+        return amounts;
+    }
+
+    /// @notice Sell the wToken of a given series to the AMM in exchange for collateral token
+    /// @param seriesId The ID of the Series to buy wToken on
+    /// @param bTokenAmount The amount of bToken to sell (bToken has the same decimals as the underlying)
+    /// @param path The path of the collateral token of the series to the collateral the caller wishes to receive
+    /// @param tokenAmountOutMinimum The lowest amount of collateral the caller is willing to receive as payment
+    /// @param sirenAmmAddress address of the amm that we wish to call
+    /// @param deadline deadline the transaction must be completed by
+    /// @param _router address of the router we wish to use ( QuickSwap or SushiSwap )
+    /// We supply a bToken and then select which collateral we wish to receive as our payment ( if it isnt the underlying asset )
+    function bTokenSell(
+        uint64 seriesId,
+        uint256 bTokenAmount,
+        address[] calldata path,
+        uint256 tokenAmountOutMinimum,
+        address sirenAmmAddress,
+        uint256 deadline,
+        address _router
+    ) external returns (uint256[] memory amounts) {
+        uint256 collateralAmountIn = ISirenTradeAMM(sirenAmmAddress)
+            .bTokenGetCollateralOut(seriesId, bTokenAmount);
+        uint256[] memory amountsOut = IUniswapV2Router02(_router).getAmountsOut(
+            collateralAmountIn,
+            path
+        );
+
+        require(
+            amountsOut[amountsOut.length - 1] >= tokenAmountOutMinimum,
+            "Minimum token ammunt out not met"
+        );
+
+        //Transfer token back to the user
+        erc1155Controller.safeTransferFrom(
+            msg.sender,
+            address(this),
+            SeriesLibrary.bTokenIndex(seriesId),
+            bTokenAmount,
+            dataReturn()
+        );
+
+        erc1155Controller.setApprovalForAll(address(sirenAmmAddress), true);
+
+        ISirenTradeAMM(sirenAmmAddress).bTokenSell(
+            seriesId,
+            bTokenAmount,
+            collateralAmountIn
+        );
+
+        TransferHelper.safeApprove(
+            path[0],
```

Comment:
Nit... it may make it more readable if these values are stored in named variables... e.g. path[0] and amountsOut[0], etc

Reply:
Gothca I was at the local variable limit STACK_TO_DEEP Error so was trying to find a way to reduce local vars, if there is a simpler way to reduce them that'd be awesome. I did try to do block scoping but it didnt seem to work when I attempted it.